examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,398 @@
+    // debug_message weight is a good approximation of the additional overhead of going from
+    // contract layer to substrate layer.
+    let overhead = <T as pallet_contracts::Config>::Schedule::get()
+        .host_fn_weights
+        .debug_message;
+    let charged_weight = env.charge_weight(base_weight.saturating_add(overhead))?;
+    trace!(
+        target: "runtime",
+        "[ChainExtension]|call|decrease_allowance / charge_weight:{:?}",
+        charged_weight
+    );
+
+    let owner = env.ext().caller();
+    let mut allowance =
+        <pallet_assets::Pallet<T> as AllowanceInspect<T::AccountId>>::allowance(
+            input.asset_id,
+            owner,
+            &input.spender,
+        );
+    <pallet_assets::Pallet<T>>::cancel_approval(
+        RawOrigin::Signed(owner.clone()).into(),
+        input.asset_id,
+        T::Lookup::unlookup(input.spender.clone()),
+    )
+    .map_err(map_err("ChainExtension failed to call decrease_allowance"))?;
+    allowance.saturating_reduce(input.value);
+    if allowance.is_zero() {
+        // If reduce value was less or equal than existing allowance, it should stay none.
+        env.adjust_weight(
+            charged_weight,
+            <T as pallet_assets::Config>::WeightInfo::cancel_approval()
+                .saturating_add(overhead),
+        );
+        return Ok(())
+    }
+    <pallet_assets::Pallet<T> as AllowanceMutate<T::AccountId>>::approve(
+        input.asset_id,
+        owner,
+        &input.spender,
+        allowance,
+    )
+    .map_err(map_err("ChainExtension failed to call decrease_allowance"))?;
+
+    trace!(
+        target: "runtime",
+        "[ChainExtension]|call|decrease_allowance"
+    );
+    Ok(())
+}
+
+impl<T> ChainExtension<T> for Psp22Extension
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+    fn call<E: Ext>(
+        func_id: u32,
+        env: Environment<E, InitState>,
+    ) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId:
+            UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+    {
+        match func_id {
+            // Note: We use the PSP22 interface selectors as function IDs,
+            // there is no need but it makes sense from a convention perspective.
+
+            // PSP22 Metadata interfaces
+            0x3d261bd4 | 0x34205be5 | 0x7271b782 => metadata::<T, E>(func_id, env)?,
+
+            // PSP22 interface queries
+            0x162df8c2 | 0x6568382f | 0x4d47d921 => query::<T, E>(func_id, env)?,
+
+            // P2P22:transfer
+            0xdb20f9f5 => transfer::<T, E>(env)?,
+
+            // P2P22:transfer_from
+            0x54b3c76e => transfer_from::<T, E>(env)?,
+
+            // PSP22::approve + PSP22::increase_allowance
+            0xb20f1bbd | 0x96d6b57a => approve::<T, E>(env)?,
```

Comment:
How come you use the same function for both?

Reply:
Good question. I'll check this with @stiiifff.

Reply:
Because of [this](https://github.com/paritytech/substrate/blob/ca1f81cd0864ae0614aaff07fa121d2d42110586/frame/assets/src/functions.rs#L722-L726). Problem is the doc on the trait function `approve` is not clear on the behavior of subsequent calls. So it's a bit of a shortcut in the context of this sample, I think it's fine .. what you think ?

Reply:
Yeah I guess that's fine. Can you add some docs around that in source code?

Reply:
OK, I've thought this through and I believe we can still implement the correct behaviour by calling `cancel_approval`, as we do in `decrease_allowance` implementation. I'd like to close this PR soon, but I'll leave a 'FIXME' comment to mark thus.